### PR TITLE
Output Type Revamp (Offline WAD Generation, Multiple Outputs)

### LINKED ISSUE #1399: ARM32 support
It would be helpful if the repo had versions of the `Compress` and `Decompress` tools compiled for 32-bit ARM (`platform.uname()[4] == 'armv7l'`) so that the randomizer could be run on a Raspberry Pi (whose 64-bit OS is [still in beta](https://www.raspberrypi.org/forums/viewtopic.php?t=275370)).

### PR DESCRIPTION
**Pull request by Cuphat**
This is a refactoring of the parts of `Main.py` responsible for outputting patch files or ROMs. The `compress_rom` setting has been split into 4 settings for outputting each of a patch file, an uncompressed ROM, a compressed ROM, or a WAD file.

To accomplish WAD generation, gzinject binaries are included for all architectures that we have Compress/Decompress binaries for. Windows binaries are 0.3.1 from the gzinject release page, Linux (x86_64) and Mac binaries are from the 0.3.6 gz releases, and Linux ARM32 and Linux ARM64 binaries were compiled using the latest source on a Raspberry Pi 3 and an ARM64 Chromebook respectively.

Additionally, added Linux ARM32 binaries for the Compressor and Decompressor, which closes #1399.

### CODE REVIEW

**Comment by RawZ06** [comment]
I guess the build of gzinject and compress are only for intel chip in macOS. 
Is it possible to build to M1 architecture too ?

**Comment by Cuphat** [comment]
The Intel Mac binaries will work through Rosetta 2. It is likely possible to build M1 binaries but without an M1 Mac I have no way to test that they would work.

**Comment by RawZ06** [comment]
I would like to buy a new M1 Mac and i don't know how the randomizer can be used on the new architecture.
Actually, I have, on my Intel Mac a virtual machine of Windows 10 to use gzinject gui version to generate wad file. 
I can try to build when I will have a new one and suggest the build here.

**Comment by Cuphat** [comment]
As far as ARM64 macOS, currently the first time you try to run a binary compiled for x86_64 macOS, it will prompt you to install Rosetta 2. Once installed, it will simply work transparently, so the current binaries, including the offline release builds, should work fine. Eventually, I am sure Apple will drop support for Rosetta 2 like they did with the original Rosetta they had when they switched from PowerPC to x86, but that is going to be a decent while down the road. That said, if anyone is interested in creating or testing ARM64 macOS binaries, I'm sure that'd be helpful.

**Comment by RawZ06** [comment]
Ok thank you for your response.

**Comment by RawZ06** [comment]
Hi, received my new mac with M1, and today I tested compress and decompress library for ARM and it works ! 
I just add it 

```
   elif platform.system() == 'Darwin':
        if platform.machine() == 'arm64':
            compressor_path += "Compress_ARM64.out"
        else:
            compressor_path += "Compress.out"
```
In main.py 

And 
```
           elif platform.system() == 'Darwin':
                if platform.machine() == 'arm64':
                    subcall = [sub_dir + "Decompress_ARM64.out", file, decomp_file]
                else:
                    subcall = [sub_dir + "Decompress.out", file, decomp_file]
```

To try it.

**Comment by Cuphat** [comment]
Alright, I managed to get gzinject to build for ARM64 macOS, so I have added that, force-pushed, and added a check into the randomizer coded to use them. Let me know if it has issues.

**Comment by RawZ06** [comment]
Thank you. I will test all, and I'll get back to you.

File: Music.py [hunk 1]
```diff
@@ -412,13 +412,16 @@ def randomize_music(rom, settings, music_mapping):
...
     if settings.patch_file != '':
         rom_version_bytes = rom.read_bytes(0x35, 3)
         rom_version = f'{rom_version_bytes[0]}.{rom_version_bytes[1]}.{rom_version_bytes[2]}'
         if compare_version(rom_version, '4.11.13') < 0:
             errors.append("Custom music is not supported by this patch version. Only randomizing vanilla music.")
             custom_sequences_enabled = False
+    elif not custom_sequences_enabled and (settings.background_music == 'random_custom_only' or settings.fanfares == 'random_custom_only'):
+        errors.append("Custom music is disabled when creating patch files. Only randomizing vanilla music.")
+
     if custom_sequences_enabled:
```
Comment (cjohnson57) [inline Music.py:425]: Apologies if I'm being dumb, but wouldn't this situation be possible?

Generating a patch file and `disable_custom_music` equals False (so custom music is enabled)

`custom_sequences_enabled` = True
 
```           
      v Not True = False, so the custom music isn't caught when it should be
elif not custom_sequences_enabled and (settings.background_music == 'random_custom_only' or settings.fanfares == 'random_custom_only'):
```

It's a little confusing with all the true/false flip flopping but I think that's right

**Review by cjohnson57 — COMMENTED** [review]
Why is it not necessary to provide a wii common key file like it is on the site?

Very minor comment: Not sure if it's possible to gray out the base ROM when all output types are disabled since normally a single choice enables/disables something. It doesn't make you provide one if there's no output anyway so it's fine either way

**Comment by Cuphat** [comment]
> Why is it not necessary to provide a wii common key file like it is on the site?

gzinject is capable of generating it.

> Very minor comment: Not sure if it's possible to gray out the base ROM when all output types are disabled since normally a single choice enables/disables something. It doesn't make you provide one if there's no output anyway so it's fine either way

Yeah, there really isn't a way to disable stuff based on the values of multiple settings as far as I know.

**Review by cjohnson57 — APPROVED** [review]

File: SettingsList.py [hunk 2]
```diff
@@ -1928,39 +1930,126 @@ def __init__(self, name, gui_text, min, max, default, step=1,
...
         gui_text       = 'Create Cosmetics Log',
+        gui_tooltip='''\
+                 Cosmetics Logs are only output if one of the output types below are enabled.
+                 ''',
         default        = True,
         disabled_default = False,
     ),
     Setting_Info(
-        name           = 'compress_rom',
+        name           = 'output_types',
```
Comment (Zannick) [inline SettingsList.py:1940]: Is this used for anything?

Comment (Cuphat) [reply SettingsList.py:1940]: Just GUI text

**Pull request merged**